HPCC-22955 Add caching of public/private key files

RussWhitehead opened the pull request:
Crypto Plugin should cache public/private key files,
since it is costly to load them every time

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ x] My code follows the code style of this project.
  - [x ] My code does not create any new warnings from compiler, build system, or lint.
- [x ] The commit message is properly formatted and free of typos.
  - [x ] The commit message title makes sense in a changelog, by itself.
  - [x ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x ] I have read the CONTRIBUTORS document.
- [x ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x ] I have given due consideration to all of the following potential concerns:
  - [x ] Scalability
  - [ x] Performance
  - [ ] Security
  - [ x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith committed:
HPCC-22869 Lightweight join must always ensure sides are ungrouped

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

richardkchapman committed:
Split off 7.2.46

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
HPCC-22926 Minor improvements to stats about expressions created

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

ghalliday committed:
HPCC-22930 Suppress location annotations when syntax checking

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

kunalaswani committed:
HPCC-22774 Translations for 7.6.X

Additional translations for ES, BS, HR, HU, PT-BR, SR, ZH added.

Signed-off-by: Kunal Aswani <Kunal.Aswani@lexisnexisrisk.com>

ghalliday committed:
HPCC-22935 Reduce the lexer overhead when skipping tokens

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

richardkchapman committed:
HPCC-21745 Prevent automatic translation from published format if reading as different file type

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Split off 7.6.2

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
HPCC-22938 Report better error for LIBRARY with function parameter

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

mayx committed:
HPCC-22937 Clear passphrase before and after signing

- Add separate function clearPassphrase
- Clear the passphrase both before and after signing

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

ghalliday committed:
HPCC-22945 Ensure that -D option works in combination with -main

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

richardkchapman committed:
HPCC-22944 Roxie incorrectly copying files to hpcc-data2 in "simple" mode

Incorrectly calculating replication level in "simple" replication mode.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

jakesmith committed:
HPCC-22946 Fix unconditional signal in end handler

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

richardkchapman committed:
Merge pull request #13053 from ghalliday/issue22938

HPCC-22938 Report better error for LIBRARY with function parameter

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #13045 from ghalliday/issue22930

HPCC-22930 Suppress location annotations when syntax checking

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #13061 from kunalaswani/HPCC-22774

HPCC-22774 Translations for 7.6.X

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #13059 from jakesmith/hpcc-22946

HPCC-22946 Fix unconditional signal in end handler

Reviewed-By: Mark Kelly <mark.kelly@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #13044 from ghalliday/issue22926

HPCC-22926 Minor improvements to stats about expressions created

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #13057 from ghalliday/issue22945

HPCC-22945 Ensure that -D option works in combination with -main

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

Commit:
HPCC-22952 Typo in Python instructions

Signed-off-by: Jim DeFabia <jamesdefabia@lexisnexis.com>

richardkchapman committed:
Merge pull request #13063 from JamesDeFabia/HPCC-22952TypoPython

HPCC-22952 Typo in Python instructions

Reviewed-By: Greg Panagiotatos <Greg.Panagiotatos@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

RussWhitehead commented:
@rpastrana Please review

HPCCSmoketest commented:
Process of PR-13064, label: keyCache762 starts now.
The reason of this test is: Did not build yet.
Commit ID: d1ab90fad4deacc57b0b7d9d873417c3b5ae00b4
Estimated completion time is ~0.5 hour(s)

jakesmith commented:
These may not be an immediate concerns, but:
+ What happens if we need to rollover the keys from time to time, i.e. load new versions?
+ Should there be a way to expunge a cached key?
+ Should there be a configurable lifetime? (e.g. if cache hit, but key older than threshold, then throws away and reloads)

jakesmith commented:
Ah ok, I had assumed for general use, and I think it should be so that others can use it too.
Esp and dafilesrv will currently be reloading the keys frequently, so it would be good to put this cache in front of those use cases too (but using the cache can be a separate PR).

RussWhitehead commented:
Opened https://track.hpccsystems.com/browse/HPCC-22957 to look into caching for Cryptohelper library.
@richardkchapman Please merge

Commit:
HPCC-22958 Std.Thorlib.Group has duplicate topic ID

Signed-off-by: Jim DeFabia <jamesdefabia@lexisnexis.com>

rpastrana commented on plugins/cryptolib/cryptolib.cpp:
```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
+public:
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)//
+    {
+        VStringBuffer searchKey("%s_%s_%s", isEmptyString(keyFS) ? "" : keyFS, isEmptyString(keyBuff) ? "" : keyBuff, isEmptyString(passphrase) ? "" : passphrase);
```
Shouldn't we bail out and avoid entering the CritBlock if keyFS and KeyBuff are empty?

RussWhitehead replied:
Done

rpastrana commented on plugins/cryptolib/cryptolib.cpp:
```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
+            {
+                //Create CLoadedKey from filespec
+                if (isPublic)
+                    pCLK = loadPublicKeyFromFile(keyFS, passphrase);
```
it seems we could load the keys outside of the critblock and we should only protect the lookup and the insert

rpastrana reviewed (changes requested):
@RussWhitehead left a few comments

RussWhitehead commented on plugins/cryptolib/cryptolib.cpp:
```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
+    KeyCache keyCache;
+
+public:
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)//
```
Good catch !

HPCCSmoketest commented:
Process of PR-13064, label: keyCache762 starts now.
The reason of this test is: New commit.
Commit ID: c182776bfcaabbbdfd33d9645c7c980724a41e38
Estimated completion time is ~0.5 hour(s)

RussWhitehead commented:
@rpastrana Please rereview

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: c182776bfcaabbbdfd33d9645c7c980724a41e38
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |864 |864 |0 |181 sec  (00:03:01) |
|test (thor) |788 |788 |0 |704 sec  (00:11:44) |
|test (roxie) |939 |939 |0 |245 sec  (00:04:05) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   41 sec (00:00:41) |  254 sec (00:04:14) |    0 sec (00:00:00) |    3 sec (00:00:03) |   21 sec (00:00:21) | 1470 sec (00:24:30) |   22 sec (00:00:22) | 1811 sec (00:30:11) |

rpastrana reviewed (approved):
@RussWhitehead approved

richardkchapman commented:
This should target 7.6.x (or possibly master), not 7.6.2

richardkchapman commented:
Checkboxes are messed up.

richardkchapman commented:
@ghalliday Thoughts? Security implications?

richardkchapman committed:
Merge pull request #13067 from JamesDeFabia/regroupLink

HPCC-22958 Std.Thorlib.Group has duplicate topic ID

Reviewed-By: Greg Panagiotatos <Greg.Panagiotatos@lexisnexis.com>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge branch 'candidate-7.2.x' into candidate-7.4.x

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge branch 'candidate-7.4.x' into candidate-7.6.x

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-22967 Poor error reporting when embedded java cannot construct row

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
Merge pull request #13058 from richardkchapman/roxie-replication

HPCC-22944 Roxie incorrectly copying files to hpcc-data2 in "simple" mode

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

jakesmith committed:
HPCC-22969 Fix bug creating empty parts on larger clusters

If multiple target clusters were specified for an output,
e.g. CLUSTER('cluster1', 'cluster2'), and if those clusters
were larger than the running cluster, every other cluster
would not have empty parts created for parts>=local # parts.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

richardkchapman committed:
Merge pull request #13073 from jakesmith/hpcc-22969-emptyparts-external-clsuter

HPCC-22969 Fix bug creating empty parts on larger clusters

Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
Merge pull request #13054 from mayx/HPCC-22937-ClearPassphrase

HPCC-22937 Clear passphrase before and after signing

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

ghalliday committed:
Merge pull request #13051 from richardkchapman/roxie-csv-as-flat

HPCC-21745 Prevent automatic translation from published format if reading as different file type

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

ghalliday committed:
Merge pull request #13071 from richardkchapman/hpcc-22967

HPCC-22967 Poor error reporting when embedded java cannot construct row

Reviewed-by: Gavin Halliday <ghalliday@hpccsystems.com>

richardkchapman committed:
Merge pull request #13056 from ghalliday/issue22935

HPCC-22935 Reduce the lexer overhead when skipping tokens

Reviewed-By: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>
Reviewed-By: Richard Chapman <rchapman@hpccsystems.com>

ghalliday commented on plugins/cryptolib/cryptolib.cpp:
```diff
@@ -570,7 +620,7 @@ CRYPTOLIB_API void CRYPTOLIB_CALL clPKIEncrypt(size32_t & __lenResult,void * & _
 {
     verifyPKIAlgorithm(pkalgorithm);
-    Owned<CLoadedKey> publicKey = loadPublicKeyFromFile(publickeyfile, passphrase);
+    Owned<CLoadedKey> publicKey = g_KeyCache.getInstance(true, publickeyfile, nullptr, passphrase);
```
Are there any threading issues with this keys being shared between multiple callers?  I would expect it to work, but double checking...

ghalliday commented on plugins/cryptolib/cryptolib.cpp:
```diff
@@ -527,6 +527,56 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
+                    pCLK = loadPrivateKeyFromMemory(keyBuff, passphrase);
+            }
+
+            keyCache.insert(pair<string, CLoadedKey*>(searchKey.str(), pCLK));
```
For efficient to use emplace, and you can use a pair of the matching type then.

ghalliday reviewed (commented):
@RussWhitehead a couple of minor comments.  Did you see any performance boost with the change in place?
